kornia/contrib/extract_patches.py

```diff
@@ -242,45 +240,26 @@ def combine_tensor_patches(
     original_size = cast(Tuple[int, int], _pair(original_size))
     window_size = cast(Tuple[int, int], _pair(window_size))
     stride = cast(Tuple[int, int], _pair(stride))
-
-    if stride[0] != window_size[0] or stride[1] != window_size[1]:
-        raise NotImplementedError(
-            f"Only stride == window_size is supported. Got {stride} and {window_size}."
-            "Please feel free to drop a PR to Kornia Github."
-        )
-
-    if unpadding:
-        unpadding = cast(PadType, _pair(unpadding))
-
-        if len(unpadding) not in [2, 4]:
-            raise AssertionError("Unpadding must be either an int, tuple of two ints or tuple of four ints")
-
-        if len(unpadding) == 2:
-            pad_vert = _pair(unpadding[0])
-            pad_horz = _pair(unpadding[1])
-        else:
-            pad_vert = unpadding[:2]
-            pad_horz = unpadding[2:]
-        unpadding = cast(Tuple[int, int, int, int], pad_horz + pad_vert)
-
-        hpad_check = (original_size[0] + unpadding[2] + unpadding[3]) % window_size[0] == 0
-        wpad_check = (original_size[1] + unpadding[0] + unpadding[1]) % window_size[1] == 0
-
-        if not hpad_check or not wpad_check:
-            raise NotImplementedError("Insufficient padding")
-
-        window_size = (
-            (original_size[0] + (unpadding[2] + unpadding[3])) // window_size[0],
-            (original_size[1] + (unpadding[0] + unpadding[1])) // window_size[1],
-        )
-
-    patches_tensor = patches.view(-1, window_size[0], window_size[1], *patches.shape[-3:])
-    restored_tensor = concatenate(torch.chunk(patches_tensor, window_size[0], 1), -2).squeeze(1)
-    restored_tensor = concatenate(torch.chunk(restored_tensor, window_size[1], 1), -1).squeeze(1)
-
-    if unpadding:
-        unpadding = cast(Tuple[int, int, int, int], unpadding)
-        restored_tensor = pad(restored_tensor, [-i for i in unpadding])
+    unpadding = cast(Tuple[int, int], _pair(unpadding))
+
+    if len(unpadding) != 2:
+        raise AssertionError("Unpadding must be either an int or a tuple of two ints")
+
+    fold = nn.Fold(output_size=original_size, kernel_size=window_size, stride=stride, padding=unpadding)
+    unfold = nn.Unfold(kernel_size=window_size, stride=stride, padding=unpadding)
+    ones = torch.ones(patches.shape[0], patches.shape[2], original_size[0], original_size[1])
+    # Assuming BNCHW
+    patches = patches.permute(0, 2, 3, 4, 1)
+    patches = patches.reshape(patches.shape[0], -1, patches.shape[-1])
+    dtype = patches.dtype
+    patches = patches.double()
```

Comment:
```suggestion
    patches = patches.to(dtype)
```
?

Reply:
This is required since `fold` does not work with other data types like (int or long). The casting to double is required for `torch.autograd.gradcheck` since it needs double precision numbers to run.

Reply:
The `torch.autograd.gradcheck` should run with just fp64, normally we use the `kornia.testing.tensor_to_gradcheck_var` (https://github.com/kornia/kornia/blob/master/kornia/testing/__init__.py#L43) to ensure it -- the casting here shouldn't be needed

Reply:
```suggestion
```

Reply:
I'm not exactly sure what you mean. The casting has to be done so that the function works with any int tensor (which is necessary for the doctests of this function to pass).  

The choice of double as an intermediate dtype was because gradcheck had raised a error (`RuntimeError: Jacobian mismatch for output 0 with respect to input 0`) when I had previously run tests. 

Do you mean have an explicit check as follows?

```python
int_flag = 0
if not torch.is_floating_point(patches):
    int_flag = 1
    dtype = patches.dtype
    patches = patches.float()

....rest..of...code

if int_flag:
    restored_tensor = restored_tensor.to(dtype)
```

Reply:
if you want to make it work for integer tensors yes -- this will avoid casting between floating types, which will save some ops

Reply:
The gradcheck test should run just within `torch.float64` tensors